challenge3/webapp/caesaris/app/services/data_series_service.rb

```diff
@@ -0,0 +1,22 @@
+class DataSeriesService
+  def date_labels
+    Dataset.select(:year, :month).group(:year, :month)
+      .order(:year, :month).map(&:date_str)
```

Comment:
`map` を使うと、ActiveRecord のオブジェクト化が必要となるので、
その処理をスキップできるように、pluck を使う方が良い。

また、この場合、group を count 等と組合わせず利用しているがこれは
あまりよくない、その場合は、distinct を使うべき。

```suggestion
      Dataset.select(:year, :month).order(:year, :month).distinct.pluck(:year, :month).map do |year, month| "#{year}-#{month}"; end
```

```

Reply:
[IMO]
この場面では`pluck` と `select`が重複しており、実際、`select(:year, :month)` の一言がなくても、発行されたSQL文が全く一緒でした。

```ruby
Dataset.order(:year, :month).distinct.pluck(:year, :month)
   (1.0ms)  SELECT DISTINCT "datasets"."year", "datasets"."month" FROM "datasets" ORDER BY "datasets"."year" ASC, "datasets"."month" ASC

Dataset.select(:year, :month).order(:year, :month).distinct.pluck(:year, :month)
   (1.3ms)  SELECT DISTINCT "datasets"."year", "datasets"."month" FROM "datasets" ORDER BY "datasets"."year" ASC, "datasets"."month" ASC
```

また、ActiveRecord のオブジェクト化を省けるのはメリットではありますが、[DRYの原則](https://ja.wikipedia.org/wiki/Don%27t_repeat_yourself)にやはり抵触しており、一応今回はただのStringの連結でよかったんですが、もし複雑なロジックの場合、将来変更しようとする時、同じロジックを記述した箇所を全部修正しなくちゃ行けなく、テストもしんどいと思います。

Reply:
下記のcommitで修復しました
https://github.com/jerrywdlee/EnergyDataSimulationChallenge/commit/f0a227b15c6886a8ef642b5a820016e425920c1c

Reply:
たしかに、select は不要でした。